src/signature.c

```diff
@@ -24,198 +24,342 @@
 
-	if (!store) {
-		return NULL;
+	fp_pubkey = fopen(CERTNAME, "r");
+	if (!fp_pubkey) {
+		fprintf(stderr, "Failed fopen %s\n", CERTNAME);
+		goto error;
 	}
-	if (X509_STORE_load_locations(store, ca_filename, ca_dirname) != 1) {
-		goto err;
+
+	cert = PEM_read_X509(fp_pubkey, NULL, NULL, NULL);
+	if (!cert) {
+		fclose(fp_pubkey);
+		goto error;
 	}
-	if (X509_STORE_set_default_paths(store) != 1) {
-		goto err;
+
+	pkey = X509_get_pubkey(cert);
+	if (!pkey) {
+		fclose(fp_pubkey);
+		X509_free(cert);
+		goto error;
 	}
-	if (crl_filename) {
-		X509_LOOKUP *lookup = X509_STORE_add_lookup(store, X509_LOOKUP_file());
-		if (!lookup) {
-			goto err;
+	return true;
+error:
+	ERR_print_errors_fp(stderr);
+	return false;
+}
+
+/* This is the main part of the signature validation.
+ * This function reads a file in chunks of 4096 bytes to create a hash from
+ * the content, then verifies the hash using the publick key against the
+ * signature.
+ *
+ * returns: true if signature was correct, false otherwise */
+static bool validate_signature(FILE *fp_data, FILE *fp_sig)
+{
+	char buffer[BUFFER_SIZE];
+	unsigned char sig_buffer[BUFFER_SIZE];
+	size_t sig_len;
+	size_t data_size;
+	size_t len = 0;
+	size_t bytes_left;
+	size_t count;
+	struct stat st;
+	int fd;
+	EVP_MD_CTX md_ctx;
+
+	sig_len = fread(sig_buffer, 1, BUFFER_SIZE, fp_sig);
+	if (sig_len == 0) {
+		fprintf(stderr, "Failed to get signature length!\n");
+		goto error;
+	}
+	fd = fileno(fp_data);
+	if (fstat(fd, &st) != 0) {
+		fprintf(stderr, "Failed to stat data file\n");
+		goto error;
+	}
+	data_size = st.st_size;
+
+	/* Initialize verify context and use sha256 algorithm internally */
+	if (!EVP_VerifyInit(&md_ctx, EVP_sha256())) {
+		goto error;
+	}
+
+	/* read all bytes from file to calculate digest using sha256 and then sign it */
+	bytes_left = data_size;
+	while (bytes_left > 0) {
+		count = (bytes_left > BUFFER_SIZE ? BUFFER_SIZE : bytes_left);
+		len = fread(buffer, 1, count, fp_data);
+		if (len != count) {
+			fprintf(stderr, "Signature check failed, read %lu bytes, expected %lu\n", len, count);
+			goto error;
 		}
-		if (X509_load_crl_file(lookup, crl_filename, X509_FILETYPE_PEM) != 1) {
-			goto err;
+		/* Hashes len bytes into the verification context */
+		if (!EVP_VerifyUpdate(&md_ctx, buffer, len)) {
+			goto error;
 		}
+		bytes_left -= len;
+	}
+
+	/* Verify data in context using pkey against signature */
+	if (EVP_VerifyFinal(&md_ctx, sig_buffer, sig_len, pkey) == 1) {
+		return true;
+	}
+
+error:
+	ERR_print_errors_fp(stderr);
+	return false;
+}
+
+/* This function makes sure the certificate is still valid by not having any
+ * compromised certificates in the chain.
+ * If there is no Certificate Revocation List (CRL) it may be that the private
+ * keys have not been compromised or the CRL has not been generated by the
+ * Certificate Authority (CA)
+ *
+ * returns: true if certificate is valid, false otherwise */
+static bool validate_certificate(void)
+{
```

Comment:
Looking at this closely, it appears we're not actually doing anything here. This is one big, noisy no-op.

Effectively we're loading a trusted cert chain of 1 cert into the store, loading a NULL revocation list, then calling X509_verify_cert() with the same cert we loaded into the store. 

Until we add in support for revocation, as well as move to a trusted cert-chain with root up around Mount Olympus, we should not make a call into this function. Keeping the code is fine, since we'll be using is eventually, but some added comments regarding when it should be called would be useful